File: python-sdk/server/configurator/configuration.py
from typing import Optional, Dict, TYPE_CHECKING

from server.configurator.blueprints.base_blueprint import BlueprintId, ModuleType

if TYPE_CHECKING:
    from server.configurator import General
    from server.configurator.blueprints import BlueprintsDB
    from server.configurator.world.player import Player
    from server.configurator.world.world import World


class Configuration:
    def __init__(self,
                 general: Optional["General"] = None,
                 blueprints: Optional["BlueprintsDB"] = None,
                 world: Optional["World"] = None,
                 players: Dict[str, "Player"] = {}):
        self.general: Optional["General"] = general
        self.blueprints: Optional["BlueprintsDB"] = blueprints
        self.world: Optional["World"] = world
        self.players: Dict[str, "Player"] = players

    def set_general(self, general: "General") -> 'Configuration':
        self.general = general
        return self

    def set_blueprints(self, blueprints: "BlueprintsDB") -> 'Configuration':
        self.blueprints = blueprints
        return self

    def add_player(self, player: "Player") -> 'Configuration':
        assert player.login not in self.players
        self.players.update({player.login: player})
        return self

    def set_world(self, world: "World") -> 'Configuration':
        self.world = world
        return self

    def verify(self):
        assert self.general
        assert self.blueprints
        assert self.world
        self.general.verify()
        self.blueprints.verify()
        self.world.verify()
        for login, player in self.players.items():
            assert len(login) > 3
            player.verify()
        self._deep_verification()

    def to_pod(self):
        self.verify()
        return {
            "application": self.general.to_pod(),
            "Blueprints": self.blueprints.to_pod(),
            "Players": {login: player.to_pod() for login, player in self.players.items()},
            "World": self.world.to_pod()
        }

    def _deep_verification(self):
        # Check, that all configured modules on a ship has an appropriate
        # blueprint. Note that not every module on the ship may be configured
        for player in self.players.values():
            for ship in player.ships.values():
                blueprint_id = BlueprintId(ModuleType.e_SHIP, ship.ship_type)
                assert blueprint_id in self.blueprints.blueprints,\
                    f"The '{blueprint_id}' blueprint doesn't exist!"

File: typescript-sdk/tests/configurator/configuration.ts
import { BlueprintId, BlueprintsDB, ModuleType } from "./blueprints.js";
import { General } from "./general.js";
import { Player, World } from "./world.js";

export interface ConfigurationOptions {
    general: General;
    blueprints: BlueprintsDB;
    world: World;
    players?: readonly Player[] | Readonly<Record<string, Player>>;
}

export class Configuration {
    general: General;
    blueprints: BlueprintsDB;
    world: World;
    readonly players = new Map<string, Player>();

    constructor(options: ConfigurationOptions) {
        this.general = options.general;
        this.blueprints = options.blueprints;
        this.world = options.world;

        const players = Array.isArray(options.players)
            ? options.players.map((player) => [player.login, player] as const)
            : Object.entries(options.players ?? {});
        for (const [login, player] of players) {
            this.addPlayer(player, login);
        }
    }

    setGeneral(general: General): this {
        this.general = general;
        return this;
    }

    setBlueprints(blueprints: BlueprintsDB): this {
        this.blueprints = blueprints;
        return this;
    }

    setWorld(world: World): this {
        this.world = world;
        return this;
    }

    addPlayer(player: Player, key = player.login): this {
        if (this.players.has(key)) {
            throw new Error(`Player '${key}' is already configured`);
        }
        this.players.set(key, player);
        return this;
    }

    verify(): void {
        this.general.verify();
        this.blueprints.verify();
        this.world.verify();

        for (const [login, player] of this.players) {
            if (login.length <= 3) {
                throw new Error("Player configuration key is too short");
            }
            player.verify();

            for (const ship of player.ships.values()) {
                const shipBlueprint = new BlueprintId(
                    ModuleType.Ship,
                    ship.shipType,
                );
                if (!this.blueprints.has(shipBlueprint)) {
                    throw new Error(
                        `Blueprint '${shipBlueprint.toPod()}' does not exist`,
                    );
                }
            }
        }
    }

    toPod(): Record<string, unknown> {
        this.verify();
        return {
            application: this.general.toPod(),
            Blueprints: this.blueprints.toPod(),
            Players: Object.fromEntries(
                [...this.players].map(([login, player]) => [
                    login,
                    player.toPod(),
                ]),
            ),
            World: this.world.toPod(),
        };
    }
}
